### ISSUE TITLE
Missing styles and walls with cutting opening.

### ISSUE THREAD

**Issue by thuytran1676** [ifcquery/ifcplusplus#250]
Run SimpleViewerExampleQt.exe 1.8, load ifc file 190516-Hudson St PG-V_Test1.ifc
Actual result: missing almost styling and walls with cutting opening. (See attached file IFC++1_8Viewer.png)
![image](https://github.com/ifcquery/ifcplusplus/assets/139486000/12acf996-d7dc-4c26-83dc-6cd17390752a)

Actual result: model should be displayed correctly like BimVision Viewer (See attached file BimVisionViewer.png)
![image](https://github.com/ifcquery/ifcplusplus/assets/139486000/fa09fc4a-7f1a-496b-a9af-63c65963aa4a)

Tried to double check using SimpleViewerExampleQt.exe 1.8, better than in 1.8 but still missing cutting opening walls and styling. (See attached file IFC++1_1Viewer.png)
![image](https://github.com/ifcquery/ifcplusplus/assets/139486000/6f7d7c31-284d-4fce-9112-359a31568816)
[190516-Hudson St PG-V_Test1.zip](https://github.com/ifcquery/ifcplusplus/files/12380500/190516-Hudson.St.PG-V_Test1.zip)
![IFC++1_8Viewer](https://github.com/ifcquery/ifcplusplus/assets/139486000/b140a916-9c42-467b-9872-c908b7742bc9)
![BimVisionViewer](https://github.com/ifcquery/ifcplusplus/assets/139486000/0538a1bf-08d7-452f-a605-9e3f75dd9583)
![IFC++1_1Viewer](https://github.com/ifcquery/ifcplusplus/assets/139486000/6bebc311-6a8e-467a-9b93-5948818e6ad1)
[190516-Hudson St PG-V_Test1.zip](https://github.com/ifcquery/ifcplusplus/files/12380506/190516-Hudson.St.PG-V_Test1.zip)
![BimVisionViewer](https://github.com/ifcquery/ifcplusplus/assets/139486000/a7763fa9-ff63-45c0-8ca5-239537f360a4)
![IFC++1_1Viewer](https://github.com/ifcquery/ifcplusplus/assets/139486000/640fa115-c5a4-45e2-853a-da3b94627910)
![IFC++1_8Viewer](https://github.com/ifcquery/ifcplusplus/assets/139486000/b2946a20-41f8-445a-9ff5-fb667d88b19d)

**Comment by ExternPointer** [comment]
I am using only parser from this library and have correct result, so this problem related to geometry generation code
<img width="759" alt="Screenshot 2023-08-18 at 18 09 48" src="https://github.com/ifcquery/ifcplusplus/assets/30197413/314689f3-29f7-427e-8abb-f3aee54ee42e">

**Comment by thuytran1676** [comment]
So that means GeometryConverter need to be fixed? because when I tried with default GeometryConverter  still have unexpected result. I mean I did not use ConverterOSG from SimpleViewerExampleQt solution.

![image](https://github.com/ifcquery/ifcplusplus/assets/139486000/21c22ca5-49e5-4e6d-8ab7-73d3c5fc536a)

**Comment by ExternPointer** [comment]
Yes. The code responsible for the CSG operations and triangulation requires special attention. It's easier to take only the parser and rework everything else, using the IFC documentation and GeometryConverter code as a cheat sheet.
Parser is really very good in this library and fast as hell, but geometry generation not so good.

**Comment by thuytran1676** [comment]
ExternPointer, can you introduce me the approach that you used to display model correctly?

**Comment by ExternPointer** [comment]
You can contact me via the email provided in the bio

**Comment by ifcquery** [comment]
There is still room for improvement for some geometry implementations.
In case it is a commercial project, I can fix and complement any geometry type.
So far, all geometry issues with customers that use the library, have been solved.
You can contact me for an offer for specific models/geomety types. Contact details are here: https://ifcquery.com/
